src/azstoragetorch/io.py

```diff
@@ -229,9 +284,74 @@ def _compute_new_position(self, offset: int, whence: int) -> int:
             return self._client.get_blob_size() + offset
         raise ValueError(f"Unsupported whence: {whence}")
 
+    def _flush(self) -> None:
+        self._check_for_stage_block_exceptions(wait=False)
+        self._flush_write_buffer()
+        self._check_for_stage_block_exceptions(wait=True)
+
+    def _flush_write_buffer(self) -> None:
+        if self._write_buffer:
+            futures = self._client.stage_blocks(memoryview(self._write_buffer))
+            self._all_stage_block_futures.extend(futures)
+            self._in_progress_stage_block_futures.extend(futures)
+            self._write_buffer = bytearray()
+
+    def _write(self, b: _SUPPORTED_WRITE_TYPES) -> int:
+        self._check_for_stage_block_exceptions(wait=False)
+        write_length = len(b)
+        self._write_buffer.extend(b)
+        if len(self._write_buffer) >= self._WRITE_BUFFER_SIZE:
+            self._flush_write_buffer()
+        self._position += write_length
+        return write_length
+
+    def _commit_blob(self) -> None:
+        self._flush()
+        block_ids = [f.result() for f in self._all_stage_block_futures]
+        self._raise_if_duplicate_block_ids(block_ids)
+        self._client.commit_block_list(block_ids)
+
+    def _raise_if_duplicate_block_ids(self, block_ids: str) -> None:
+        # An additional safety measure to ensure we never reuse block IDs within a BlobIO instance. This
+        # should not be an issue with UUID4 for block IDs, but that may not always be the case if
+        # block ID generation changes in the future.
+        if len(block_ids) != len(set(block_ids)):
+            raise RuntimeError(
+                "Unexpected duplicate block IDs detected. Not committing blob."
+            )
+
+    def _check_for_stage_block_exceptions(self, wait: bool = True) -> None:
+        # Before doing any additional processing, raise if an exception has already
+        # been processed especially if it is going to require us to wait for all
+        # in-progress futures to complete.
+        self._raise_if_fatal_write_error()
+        self._process_stage_block_futures_for_errors(wait)
+
+    def _process_stage_block_futures_for_errors(self, wait: bool) -> None:
+        if wait:
+            concurrent.futures.wait(
+                self._in_progress_stage_block_futures,
+                return_when=concurrent.futures.FIRST_EXCEPTION,
+            )
+        futures_still_in_progress = []
+        for future in self._in_progress_stage_block_futures:
```

Comment:
So essentially this is saving the FIRST exception? Do we care about the others? I guess maybe not.

Reply:
Yep only the first one is saved. I mainly chose to do that because I'm not sure how valuable is to have all of the exceptions and the error message it could potentially get overwhelming if all simultaneous stage blocks hit the same error (e.g. permissions error) and we displayed them all. Thoughts?

Reply:
Ok, yeah, I think that makes sense and agree showing all exceptions may be a bit too much information as they are likely all the same. I also don't know a great way to show them if we don't have something like AggregateException in .NET. This is probably fine for now and if we get some feedback on it, we can adjust.

Reply:
Sounds good. Yeah I agree if we get feedback that customers want all of the exceptions, we should be able to add it. Just need to make sure the end message is still readable.